drivers/serial/uart_nrfx_uart.c

```diff
@@ -229,6 +228,11 @@ static void uart_nrfx_poll_out(struct device *dev,
 	 * blue moon against hanging up the whole thread permanently
 	 */
 
+#ifdef CONFIG_UART_0_ASYNC
+	/* Disable TXDRDY interrupts, so that ISR doesn't clear TXDRDY event. */
+	nrf_uart_int_disable(uart0_addr, NRF_UART_INT_MASK_TXDRDY);
```

Comment:
Disabling the interrupt will stop the thread locking up, at the expense of dropping any data remaining in the ASYNC send. I also don't think this fixes two threads calling poll_out at the same time, they just both disable the interrupt and still both loop on TXDRDY at the same time.

Reply:
Function is working fine with two threads calling poll_out at the same time, it is possible to lose some bytes but there is no other solution if we want to keep it lockless.
If writing from multiple contexts is desired, tty api should be used.
Not all data is dropped, it depends on timings but most of the data is sent out from async send.

Reply:
As far as I can tell, there is no way to force the logger backend or printk to route through the TTY system. If this is true, there is no way to reliably send data via UART, as printk/logger can blurt over the tty'd TX at any point in time. I understand the desire to keep it lockless, but it seems to come at the cost of making UART unusable for data protocols, rather than purely debug output.

Reply:
This doesn't seem to be the case for the other UART drivers I have scanned through, which have a dedicated TXBUSY register or similar.  If a lock is required to have this driver conform to the majority, that seems a worthy trade to me.